### PR TITLE
[untested] AWQ

### PR DESCRIPTION
**Pull request by flozi00**
This PR adds AWQ as quantization method
At the moment Mistral and Llama are supported

### CODE REVIEW

**Comment by flozi00** [comment]
Thanks a Lot
Could you please push the image somewhere if there are any bugs or errors ?
So I would use that image to fix them instead of my dev env.
I recognized some python caching issues, so sometimes updated code won't be executed and any cached version will be instead.

File: server/Makefile-awq [hunk 1]
```diff
@@ -0,0 +1,13 @@
+awq_commit := 19a5a2c9db47f69a2851c83fea90f81ed49269ab
```
Comment (tgaddair) [inline server/Makefile-awq:1]: Do you need to include this in the Dockerfile?

**Comment by tgaddair** [comment]
Hey @flozi00 looks like there was an error here:

```
Traceback (most recent call last):                                                                                                                                        [90/526]
  File "/data/lorax/server/lorax_server/models/flash_causal_lm.py", line 827, in warmup                                                                                           
    _, batch = self.generate_token(batch)                                                                                                                                         
  File "/opt/conda/lib/python3.9/contextlib.py", line 79, in inner                                                                                                                
    return func(*args, **kwds)                                                                                                                                                    
  File "/data/lorax/server/lorax_server/models/flash_causal_lm.py", line 923, in generate_token                                                                                   
    raise e                                                                                                                                                                       
  File "/data/lorax/server/lorax_server/models/flash_causal_lm.py", line 920, in generate_token                                                                                   
    out = self.forward(batch, adapter_data)                                                                                                                                       
  File "/data/lorax/server/lorax_server/models/flash_mistral.py", line 399, in forward                                                                                            
    logits = self.model.forward(                                                                                                                                                  
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 564, in forward                                                                   
    hidden_states = self.model(                                                                                                                                                   
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl                                                                              
    return forward_call(*args, **kwargs)                                                                                                                                          
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 507, in forward                                                                   
    hidden_states, residual = layer(                                                                                                                                              
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl                                                                              
    return forward_call(*args, **kwargs)                                                                                                                                          
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 430, in forward
    attn_output = self.self_attn(
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/lorax/server/lorax_server/models/custom_modeling/flash_mistral_modeling.py", line 292, in forward
    qkv = self.query_key_value(hidden_states, adapter_data)
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/lorax/server/lorax_server/utils/layers.py", line 440, in forward
    result = self.base_layer(input)
  File "/opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/lorax/server/lorax_server/utils/layers.py", line 253, in forward
    return self.linear.forward(x)
  File "/opt/conda/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/lorax/server/lorax_server/utils/awq/awq.py", line 32, in forward
    out = awq_inference_engine.gemm_forward_cuda(x.reshape(-1, x.shape[-1]), self.qweight, self.scales, self.qzeros, 8)
TypeError: gemm_forward_cuda(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch.Tensor, arg1: torch.Tensor, arg2: torch.Tensor, arg3: torch.Tensor, arg4: int, arg5: int) -> torch.Tensor

Invoked with: tensor([[-2.5868e-04,  3.9029e-04, -2.3377e-04,  ..., -5.0497e-04,                                                                                          [50/526]
          3.7956e-04,  1.8132e-04],                                                                                                                                               
        [-2.9159e-04,  6.6032e-03,  1.7571e-04,  ..., -2.1858e-03,                                                                                                                
          3.8457e-04,  2.3687e-04],                                                                                                                                               
        [ 1.2022e-04,  7.9870e-04, -2.9302e-04,  ..., -1.5945e-03,                                                                                                                
         -6.4909e-05, -1.3475e-03],                                                                                                                                               
        ...,                                                                                                                                                                      
        [-2.9159e-04,  6.6032e-03,  1.7571e-04,  ..., -2.1858e-03,                                                                                                                
          3.8457e-04,  2.3687e-04],                                                                                                                                               
        [ 1.2022e-04,  7.9870e-04, -2.9302e-04,  ..., -1.5945e-03,                                                                                                                
         -6.4909e-05, -1.3475e-03],                                                                                                                                               
        [-7.7069e-05, -1.3266e-03,  1.6689e-06,  ...,  1.0653e-03,                                                                                                                
          1.1867e-04,  2.0065e-03]], device='cuda:0', dtype=torch.float16), tensor([[  556552743, -1299357422,  1531815145,  ..., -2004322455,                                    
          2056820870,  2056755096],                                                                                                                                               
        [  556552743, -1299357422,  1531815145,  ..., -1751611542,                                                                                                                
         -1952933770,  1788315288],                                                                                                                                               
        [  556552743, -1299357422,  1531815145,  ..., -2004322455,                                                                                                                
          2056820870,  2056755096],                                                                                                                                               
        ...,
        [ 1751758709,  1984656729,  1784329302,  ..., -1703257480,
          1738168454,  1468565657],
        [ 1752807284,  1984656729,  1784329302,  ..., -1938077063,
          1737054375,  1451783833],
        [ 1752807284,  1984656729,  1784329302,  ..., -1938077063,
          1736009912,  1450800793]], device='cuda:0', dtype=torch.int32), tensor([[0.0243, 0.0618, 0.1774,  ..., 0.0124, 0.0089, 0.0146],
        [0.0363, 0.0358, 0.1422,  ..., 0.0143, 0.0175, 0.0214],
        [0.0379, 0.0304, 0.0851,  ..., 0.0177, 0.0114, 0.0140],
        ...,
        [0.0327, 0.0375, 0.0938,  ..., 0.0104, 0.0140, 0.0127],
        [0.0315, 0.0977, 0.1758,  ..., 0.0135, 0.0110, 0.0192],
        [0.0348, 0.0269, 0.0462,  ..., 0.0144, 0.0132, 0.0169]],
       device='cuda:0', dtype=torch.float16), tensor([[  556552743, -1299357422,  1531815145,  ..., -2004322455,
          2056820870,  2056755096],
        [ 1128743755, -1568886220,  1028093113,  ...,  1736935846,
          2020252058, -1755735927],
        [-1736914540,   676862297, -1033058437,  ...,  1739032922,
          2052577126, -2021102759],
        ...,
        [ 1718052440,  1721267814, -2005297478,  ..., -2051705226,
         -1752738423, -1735751529],
        [  573391397, -1551257054,   777742284,  ...,  2055645594,
         -2007406698,  1486456453],
        [ 1752807284,  1984656729,  1784329302,  ..., -1938077063,
          1737054375,  1451783833]], device='cuda:0', dtype=torch.int32), 8
```

I'll see about uploading a Docker image, we should be able to build one from this PR directly if we move it to the main `predibase/lorax` repo.

**Comment by tgaddair** [comment]
Looking at the code, it looks like the implementation of `awq.py` is referencing a different commit from the one in the Makefile. Likely we need to use the implementation from https://github.com/mit-han-lab/llm-awq/blob/19a5a2c9db47f69a2851c83fea90f81ed49269ab/awq/quantize/qmodule.py

**Comment by flozi00** [comment]
Copied the relevant changes from main branch.
Hopefully there is no syntax error by using mobile phone.

Otherwise I will check again some to go

**Pull request merged**